### PR TITLE
Add options to qos 1 subscription

### PR DESCRIPTION
**Pull request by dumbpreacher**
## Proposed Changes

This PR introduces two options to qos 1 topics in subscription in MQTT v4. The options are non_retry and non_persistence. Broker will send non_retry messages only once. Broker will not store non_persistence messages across sessions.  

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #XXXX)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CODE_OF_CONDUCT.md` document
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if needed)
- [x] Any dependent changes have been merged and published in related repositories
- [ ] I have updated changelog (At the bottom of the release version)
- [ ] I have squashed all my commits into one before merging

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

File: apps/vmq_server/src/vmq_mqtt_fsm.erl [hunk 1]
```diff
@@ -373,21 +373,25 @@ connected(#mqtt_pubcomp{message_id=MessageId}, State) ->
...
+    SubTopics = subtopics(Topics, ProtoVer),
     OnAuthSuccess =
         fun(_User, _SubscriberId, MaybeChangedTopics) ->
-                SubTopics = subtopics(MaybeChangedTopics, ProtoVer),
-                case vmq_reg:subscribe(CAPSettings#cap_settings.allow_subscribe, SubscriberId, SubTopics) of
+                case vmq_reg:subscribe(CAPSettings#cap_settings.allow_subscribe, SubscriberId, MaybeChangedTopics) of
                     {ok, _} = Res ->
-                        vmq_plugin:all(on_subscribe, [User, SubscriberId, MaybeChangedTopics]),
+                        T = lists:foldr(fun({Topic, Sub} ,Acc) -> [{Topic, extract_qos(Sub)}|Acc] end, [], MaybeChangedTopics),
+                        lager:error("Topics passed to hook ~p", [T]),
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_fsm.erl:386]: Unwanted log

File: apps/vmq_server/src/vmq_mqtt_fsm.erl [hunk 2]
```diff
@@ -373,21 +373,25 @@ connected(#mqtt_pubcomp{message_id=MessageId}, State) ->
...
+                        vmq_plugin:all(on_subscribe, [User, SubscriberId, T]),
                         Res;
                     Res -> Res
                 end
         end,
-    case auth_on_subscribe(User, SubscriberId, Topics, OnAuthSuccess) of
+    case auth_on_subscribe(User, SubscriberId, SubTopics, OnAuthSuccess) of
         {ok, QoSs} ->
+            lager:error("L394: QoSTable passed to hook ~p", [QoSs]),
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_fsm.erl:394]: Unwanted log

File: apps/vmq_server/src/vmq_mqtt_fsm.erl [hunk 3]
```diff
@@ -676,7 +680,7 @@ set_sock_opts(Opts) ->
     self() ! {set_sock_opts, Opts}.
 
 -spec auth_on_subscribe(username(), subscriber_id(), [{topic(), qos()}],
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_fsm.erl:682]:
```suggestion
-spec auth_on_subscribe(username(), subscriber_id(), [{topic(), subinfo()}],
```

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

File: apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+qos1_persist_test(Cfg) ->
+  Connect = packet:gen_connect(vmq_cth:ustr(Cfg) ++ "qos1-persist-test", [{keepalive,60}, {clean_session, false}]),
+  Connack1 = packet:gen_connack(0),
+  Connack2 = packet:gen_connack(true, 0),
+  Disconnect = packet:gen_disconnect(),
+  Subscribe = packet:gen_subscribe(109, "qos1/persistence/test", 1, false, false),
+  Suback = packet:gen_suback(109, 1),
+  Publish = packet:gen_publish("qos1/persistence/test", 1, <<"persist-message">>, [{mid, 1}]),
+  {ok, Socket} = packet:do_client_connect(Connect, Connack1, []),
+  enable_on_publish(),
```
Comment (dhruvjain99) [inline apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl:62]: This should be auth_on_publish and same for auth_on_subscribe.

File: apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  ok = gen_tcp:send(Socket, Subscribe),
+  ok = packet:expect_packet(Socket, "suback", Suback),
+
+  %% we publish message and expect that broker does not retry even if puback is not sent
+  qos1_opts_helper(packet:gen_publish("qos1/retry/test", 1, <<"retry-message">>, [{mid, 128}]), 128),
+
+  %% we receive the message but do not send puback
+  ok = packet:expect_packet(Socket, "publish", Publish),
+
+  %% wait for 25 seconds and check that we receive no further messages
```
Comment (dhruvjain99) [inline apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl:129]: This is a retry test. We expect message to be retried.

File: apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  {error,timeout} = packet:expect_packet(gen_tcp, Socket1, "publish", Publish, 500),
+  disable_on_publish(),
+  disable_on_subscribe(),
+  ok = gen_tcp:close(Socket1).
+
+qos1_retry_test(Cfg) ->
+  Connect = packet:gen_connect(vmq_cth:ustr(Cfg) ++ "qos1-retry-test", [{keepalive,60}, {clean_session, false}]),
+  Connack1 = packet:gen_connack(0),
+  Disconnect = packet:gen_disconnect(),
+  Subscribe = packet:gen_subscribe(109, "qos1/retry/test", 1, true, false),
```
Comment (dhruvjain99) [inline apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl:114]: Retry test -> non_retry should be false.

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

**Comment by dhruvjain99** [comment]
Can you please update the PR description?

File: apps/vmq_server/src/vmq_reg.erl [hunk 7]
```diff
@@ -690,9 +696,24 @@ maybe_set_expiry_ts(_) ->
...
+    Msg#vmq_msg{non_retry = NonRetry};
+handle_retry_flag(_SubInfo, Msg) ->
+    %% Default is to set the non_retry flag to false so that default qos 1 behaviour is preserved
+    Msg#vmq_msg{non_retry = false}.
+
+-spec handle_non_persistence(subinfo(), msg()) -> msg().
+handle_non_persistence({_QoS, #{non_persistence := NonPersistence}}, Msg) ->
+    Msg#vmq_msg{non_persistence = NonPersistence};
+handle_non_persistence(_SubInfo, Msg) ->
+    %% Default is to set the retain flag to false to be compatible with MQTTv3
```
[the remark this answers is not in the document]
Comment (dumbpreacher) [reply apps/vmq_server/src/vmq_reg.erl:714]: This tells why we are setting this to false as default

File: apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl (hunk 4 above)
Comment (dumbpreacher) [reply apps/vmq_server/test/vmq_mqtt_opts_SUITE.erl:62]: Why?

**Pull request merged**